File: kotlin-native/backend.native/tests/objcexport/swiftNameMangling.kt
@file:OptIn(kotlin.experimental.ExperimentalObjCName::class)

package swiftNameMangling

interface SwiftNameManglingI1 {
    val clashingProperty: Int

    fun clashingMethod(): Int

    @ObjCName(swiftName = "swiftClashingMethodWithObjCNameInI1")
    fun clashingMethodWithObjCNameInI1(): Int

    fun swiftClashingMethodWithObjCNameInI2(): Int

    @ObjCName(swiftName = "swiftClashingMethodWithObjCNameInBoth")
    fun clashingMethodWithObjCNameInBoth(): Int
}

fun i1() = object : SwiftNameManglingI1 {
    override val clashingProperty: Int
        get() = 1

    override fun clashingMethod(): Int = 2

    override fun clashingMethodWithObjCNameInI1(): Int = 3

    override fun swiftClashingMethodWithObjCNameInI2(): Int = 4

    override fun clashingMethodWithObjCNameInBoth(): Int = 5
}

interface SwiftNameManglingI2 {
    val clashingProperty: Any

    fun clashingMethod(): Any

    fun swiftClashingMethodWithObjCNameInI1(): Any

    @ObjCName(swiftName = "swiftClashingMethodWithObjCNameInI2")
    fun clashingMethodWithObjCNameInI2(): Any

    @ObjCName(swiftName = "swiftClashingMethodWithObjCNameInBoth")
    fun clashingMethodWithObjCNameInBoth(): Any
}

fun i2() = object : SwiftNameManglingI2 {
    override val clashingProperty: Any
        get() = "one"

    override fun clashingMethod(): Any = "two"

    override fun swiftClashingMethodWithObjCNameInI1(): Any = "three"

    override fun clashingMethodWithObjCNameInI2(): Any = "four"

    override fun clashingMethodWithObjCNameInBoth(): Any = "five"
}

File: kotlin-native/backend.native/tests/objcexport/swiftNameMangling.swift
import Kt

private func test1() throws {
    let i1 = SwiftNameManglingKt.i1()
    let i2 = SwiftNameManglingKt.i2()

#if DISABLE_MEMBER_NAME_MANGLING || DISABLE_INTERFACE_METHOD_NAME_MANGLING
    try assertEquals(actual: i1.clashingProperty, expected: 1)
    try assertEquals(actual: i1.clashingMethod(), expected: 2)
    try assertEquals(actual: i1.swiftClashingMethodWithObjCNameInI1(), expected: 3)
    try assertEquals(actual: i1.swiftClashingMethodWithObjCNameInI2(), expected: 4)
    try assertEquals(actual: i1.swiftClashingMethodWithObjCNameInBoth(), expected: 5)

    try assertEquals(actual: i2.clashingProperty as! String, expected: "one")
    try assertEquals(actual: i2.clashingMethod() as! String, expected: "two")
    try assertEquals(actual: i2.swiftClashingMethodWithObjCNameInI1() as! String, expected: "three")
    try assertEquals(actual: i2.swiftClashingMethodWithObjCNameInI2() as! String, expected: "four")
    try assertEquals(actual: i2.swiftClashingMethodWithObjCNameInBoth() as! String, expected: "five")
#endif
}

class SwiftNameManglingTests : SimpleTestProvider {
    override init() {
        super.init()

        test("Test1", test1)
    }
}